legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabTokenManager.java

```diff
@@ -50,169 +46,107 @@ public boolean equals(Object other)
+        GitLabToken oldToken = this.token;
+        this.token = token;
+        return !token.equals(oldToken);
     }
 
-    boolean gitLabOAuthCallback(GitLabMode mode, String code)
+    public StringBuilder appendGitLabTokenInfo(StringBuilder builder)
     {
-        GitLabModeInfo modeInfo = getModeInfo(mode);
-        if (modeInfo == null)
-        {
-            throw new IllegalStateException("Unsupported GitLab mode: " + mode);
-        }
-        synchronized (this.tokens)
-        {
-            GitLabToken token = GitLabToken.newGitLabToken(TokenType.OAUTH2_ACCESS, GitLabOAuthAuthenticator.newAuthenticator(modeInfo).getOAuthTokenFromAuthCode(code));
-            GitLabToken oldToken = this.tokens.put(mode, token);
-            return !token.equals(oldToken);
-        }
+        return builder.append("token=[").append(this.token != null ? this.token.toString() : "").append("]");
     }
 
-    StringBuilder appendTokenInfo(StringBuilder builder)
+    Token.TokenBuilder encodeGitLabToken(Token.TokenBuilder builder)
     {
-        synchronized (this.tokens)
+        builder.putInt(this.token != null ? 1 : 0);
+        if (this.token != null)
         {
-            builder.append("tokens=[");
-            int startLen = builder.length();
-            for (GitLabMode mode : GitLabMode.values())
-            {
-                GitLabToken token = this.tokens.get(mode);
-                if (token != null)
-                {
-                    ((builder.length() == startLen) ? builder : builder.append(", ")).append(mode.name()).append('=').append(token.toString());
-                }
-            }
-            builder.append(']');
-            return builder;
+            builder.putString(this.appInfo.getAppId());
+            builder.putString(this.token.getTokenType().toString());
+            builder.putString(this.token.getToken());
         }
+        return builder;
     }
 
-    Token.TokenBuilder encode(Token.TokenBuilder builder)
+    void decodeAndSetToken(Token.TokenReader reader)
     {
-        synchronized (this.tokens)
+        int tokenCount = reader.getInt();
+        if (tokenCount == 0)
         {
-            builder.putInt(this.tokens.size());
-            for (GitLabMode mode : GitLabMode.values())
-            {
-                GitLabToken token = this.tokens.get(mode);
-                if (token != null)
-                {
-                    builder.putString(mode.name());
-                    builder.putString(this.modeInfos.getModeInfo(mode).getAppInfo().getAppId());
-                    builder.putString(token.getTokenType().toString());
-                    builder.putString(token.getToken());
-                }
-            }
-            return builder;
+            return;
+        }
+        if (tokenCount > 1 || tokenCount < 0)
+        {
+            throw new IllegalStateException("Unexpected value (" + tokenCount + ") for token count");
```

Comment:
I would say that you should loop through all the tokens, regardless of how many there are. If there are extra tokens in the reader that don't match the appId, they can simply be discarded. They aren't harmful. (It's important, though, that you read through all tokens from the reader, so that it's in an appropriate state for whatever reads from it next.)